src/firstboot/firstboot.c

```diff
@@ -581,24 +603,31 @@ static int process_hostname(void) {
         if (isempty(arg_hostname))
                 return 0;
 
-        r = write_string_file(etc_hostname, arg_hostname,
-                              WRITE_STRING_FILE_CREATE | WRITE_STRING_FILE_SYNC | WRITE_STRING_FILE_MKDIR_0755 |
-                              (arg_force ? WRITE_STRING_FILE_ATOMIC : 0));
+        r = write_string_file_at(pfd, "hostname", arg_hostname,
+                                 WRITE_STRING_FILE_CREATE | WRITE_STRING_FILE_SYNC |
+                                 (arg_force ? WRITE_STRING_FILE_ATOMIC : 0));
         if (r < 0)
-                return log_error_errno(r, "Failed to write %s: %m", etc_hostname);
+                return log_error_errno(r, "Failed to write %s/%s/hostname: %m", format_fd_path(pfd, &buf), p);
 
-        log_info("%s written.", etc_hostname);
+        log_info("%s/%s/hostname written.", format_fd_path(rfd, &buf), p);
         return 0;
 }
 
-static int process_machine_id(void) {
-        const char *etc_machine_id;
+static int process_machine_id(int rfd) {
+        _cleanup_free_ char *p = NULL, *buf = NULL;
+        _cleanup_close_ int pfd = -EBADF;
         int r;
 
-        etc_machine_id = prefix_roota(arg_root, "/etc/machine-id");
-        if (laccess(etc_machine_id, F_OK) >= 0 && !arg_force) {
-                log_debug("Found %s, assuming machine-id has been configured.",
-                          etc_machine_id);
+        pfd = chase_symlinks_at_and_open(rfd, "etc",
+                                         CHASE_AT_RESOLVE_IN_ROOT|CHASE_MKDIR_0755,
+                                         O_CLOEXEC|O_PATH,
+                                         &p);
+        if (pfd < 0)
+                return log_error_errno(pfd, "Failed to create %s/etc: %m", format_fd_path(rfd, &buf));
```

Comment:
Shouldn't this be split out to a helper like `acquire_etc_fd_or_warn()`? I repeats quite a few times.